[Refactor] Refactor move phase and add documentation

DayKev opened the pull request:
Text mostly copied from #3064

## What are the changes?
~~Nothing functional, at least not intentionally.~~  Added documentation and cleaned up `MovePhase`.
Update: A bug introduced by the `strict-null` refactor (see [here](https://github.com/pagefaultgames/pokerogue/pull/3259/files#diff-7bfd86f885f7bd68a0b3b3a290cbfe482a6338ac569b6a7418716a863006537aR2657-R2658)) is fixed (multi-hit moves called by Metronome/etc only hit once).

## Why am I doing these changes?
It's extremely difficult to follow in its current state.  Adding documentation and cleaning up the code here ~~(and soon `MoveEffectPhase`)~~ will pay off in the short term, and will ultimately make it easier to game-plan a larger refactor.

## What did change?
Large sections of code were split out into methods and variables were renamed for readability; quite a few comments were added in places where I felt they'd be helpful.  A few things were shuffled around, but only in non-functional ways (e.g. moving a variable declaration closer to where it ends up being used).

## Checklist
- [x] **I'm using `beta` as my base branch**
- [ ] There is no overlap with another PR?
  - Update of #3064
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- ~[ ] Have I considered writing automated tests for the issue?~
- ~[ ] If I have text, did I make it translatable and add a key in the English locale file(s)?~
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- ~[ ] Are the changes visual?~
  - ~[ ] Have I provided screenshots/videos of the changes?~

DayKev commented on src/phases/move-phase.ts:
```diff
@@ -63,186 +65,69 @@ export class MovePhase extends BattlePhase {
+    // Reset hit-related turn data when starting follow-up moves (e.g. Metronomed moves, Dancer repeats)
+    if (this.followUp) {
+      this.pokemon.turnData.hitsLeft = 0;
+      this.pokemon.turnData.hitCount = 0;
```
Good point, that's another bug down.

flx-sta commented:
@DayKev I know this is a mirror but refactors like this unfortunately require some test coverage to make sure everything works as expected.

There is also 2 more nits:
- Check if all the public methods & fields/attributes really have to be public
- If a method is public it should have tsdoc `/**` on it
- If a method overrides it should be tagged with `override` in front

DayKev commented:
> @DayKev I know this is a mirror but refactors like this unfortunately require some test coverage to make sure everything works as expected.

That's what the test suite is for, we have plenty of tests that check a bunch of things about various moves; what would I even add that isn't already covered by another test?

> There is also some general things to check:
> 
> * Check if all the public methods & fields/attributes really have to be public
> * If a method is public it should have tsdoc `/**` on it (Same maybe for a public field/attribute)
> * If a method overrides it should be tagged with `override` in front

* I modified some of the fields from `public` to `protected` `set`/`public` `get`.
* I moved a bit of `//` comments out into tsdocs where it seemed to make sense, I think the rest of the comments are good.
* No methods are overridden.

flx-sta commented on src/battle-scene.ts:
```diff
@@ -25,7 +25,7 @@ import {
 } from "./modifier/modifier-type";
 import AbilityBar from "./ui/ability-bar";
-import { BlockItemTheftAbAttr, DoubleBattleChanceAbAttr, ChangeMovePriorityAbAttr, PostBattleInitAbAttr, applyAbAttrs, applyPostBattleInitAbAttrs } from "./data/ability";
+import { BlockItemTheftAbAttr, DoubleBattleChanceAbAttr, PostBattleInitAbAttr, applyAbAttrs, applyPostBattleInitAbAttrs } from "./data/ability";
```
NIT:
Might wanna use the `#app` import pattern

flx-sta commented on src/field/arena.ts:
```diff
@@ -391,16 +391,16 @@ export class Arena {
 
-  isMoveWeatherCancelled(user: Pokemon, move: Move) {
-    return this.weather && !this.weather.isEffectSuppressed(this.scene) && this.weather.isMoveWeatherCancelled(user, move);
+  isMoveWeatherCancelled(user: Pokemon, move: Move): boolean {
```
NIT:
- please add a scope (public/protected/private)
- please add docs

DayKev replied:
I don't see the point of docs when they would just restate the name of the function in a couple more words. Everything about the function is self-explanatory.

flx-sta replied:
It's mostly to describe the params a little better.
I know it's self explanatory but i ain't making the rules

flx-sta commented on src/field/arena.ts:
```diff
@@ -391,16 +391,16 @@ export class Arena {
 
-  isMoveTerrainCancelled(user: Pokemon, targets: BattlerIndex[], move: Move) {
-    return this.terrain && this.terrain.isMoveTerrainCancelled(user, targets, move);
+  isMoveTerrainCancelled(user: Pokemon, targets: BattlerIndex[], move: Move): boolean {
```
NIT:
- please add a scope (public/protected/private)
- please add docs

flx-sta commented on src/field/arena.ts:
```diff
@@ -391,16 +391,16 @@ export class Arena {
 
-  getTerrainType() : TerrainType {
-    return this.terrain?.terrainType || TerrainType.NONE;
+  getTerrainType(): TerrainType {
```
NIT:
- please add a scope (public/protected/private)
- please add docs

flx-sta commented on src/field/pokemon.ts:
```diff
@@ -4863,8 +4863,8 @@ export class PokemonBattleSummonData {
-  public hitCount: number;
-  public hitsLeft: number;
+  public hitCount: number = 0;
+  public hitsLeft: number = -1;
```
Q:
Why `-1` here?
I mean `0` also means there is no more hits left, right?

DayKev replied:
`0` makes the move assume it's finished and so multi-hit moves don't do all their hits, `-1` means it should calculate how many hits are left.

flx-sta replied:
@DayKev Would you mind adding that as a comment on top of this?
```suggestion
  /** 0 = Move is finished. -1 = flagged to calculate how many hits are left */
  public hitsLeft: number = -1;
```

flx-sta commented on src/phases/move-phase.ts:
```diff
@@ -40,10 +68,13 @@ export class MovePhase extends BattlePhase {
+   * @param ignoreDisableTags `true` to not check if the move is disabled
+   * @returns `true` if all the checks pass
+   */
   canMove(ignoreDisableTags?: boolean): boolean {
```
NIT:
- please add a scope (public/protected/private)

flx-sta commented on src/phases/move-phase.ts:
```diff
@@ -63,188 +94,67 @@ export class MovePhase extends BattlePhase {
-      }
-      this.end();
-    };
+  getActiveTargetPokemon() {
```
NIT:
- please add a scope (public/protected/private)
- add docs

flx-sta commented on src/phases/move-phase.ts:
```diff
@@ -63,188 +94,67 @@ export class MovePhase extends BattlePhase {
+  }
 
+  /** Handles Sleep/Paralysis/Freeze rolls and side effects, along with lapsing volatile statuses. */
+  resolvePreMoveStatusEffects() {
```
NIT:
- please add a scope (public/protected/private)
- please fix docs formatting

flx-sta commented on src/phases/move-phase.ts:
```diff
@@ -273,25 +183,260 @@ export class MovePhase extends BattlePhase {
+    }
+  }
+
+  useMove() {
```
NIT:
- please add a scope (public/protected/private)
- add docs

flx-sta commented on src/phases/move-phase.ts:
```diff
@@ -273,25 +183,260 @@ export class MovePhase extends BattlePhase {
+   * - Move redirection abilities, effects, etc.
+   * - Counterattacks, which pass a special value into the `targets` constructor param (`[BattlerIndex.ATTACKER]`).
+   */
+  resolveRedirectTarget() {
```
NIT:
- please add a scope (public/protected/private)
- add docs

flx-sta commented on src/phases/move-phase.ts:
```diff
@@ -273,25 +183,260 @@ export class MovePhase extends BattlePhase {
+   *
+   * TODO: This hardcodes the PP increase at 1 per opponent, rather than deferring to the ability.
+   */
+  getPpIncreaseFromPressure(targets: Pokemon[]) {
```
NIT:
- please add a scope (public/protected/private)
- add docs

flx-sta commented on src/phases/move-phase.ts:
```diff
@@ -273,25 +183,260 @@ export class MovePhase extends BattlePhase {
+    }
+  }
+
+  end() {
```
NIT:
- please add a scope (public/protected/private)
- add docs

flx-sta commented on src/phases/move-phase.ts:
```diff
@@ -273,25 +183,260 @@ export class MovePhase extends BattlePhase {
+   * If there is no last attacker, or they are no longer on the field, a message is displayed and the
+   * move is marked for failure.
+   */
+  resolveCounterAttackTarget() {
```
NIT:
- please add a scope (public/protected/private)

flx-sta commented on src/phases/move-phase.ts:
```diff
@@ -273,25 +183,260 @@ export class MovePhase extends BattlePhase {
+   *
+   *   TODO: handle charge moves more gracefully
+   */
+  handlePreMoveFailures() {
```
NIT:
- please add a scope (public/protected/private)

flx-sta commented on src/phases/move-phase.ts:
```diff
@@ -273,25 +183,260 @@ export class MovePhase extends BattlePhase {
+    }
   }
 
   showMoveText(): void {
```
NIT:
- please add a scope (public/protected/private)
- add docs

flx-sta commented on src/phases/move-phase.ts:
```diff
@@ -311,18 +456,10 @@ export class MovePhase extends BattlePhase {
-    }
-
-    super.end();
+  showFailedText(failedText?: string): void {
```
NIT:
- please add a scope (public/protected/private)
- add docs

flx-sta reviewed (changes requested):
Make sure to double check for potential `override` usages

DayKev commented on src/phases/move-phase.ts:
```diff
@@ -273,25 +183,260 @@ export class MovePhase extends BattlePhase {
+  lapsePreMoveAndMoveTags() {
+    this.pokemon.lapseTags(BattlerTagLapseType.PRE_MOVE);
+
+    // TODO: does this intentionally happen before the no targets/Moves.NONE on queue cancellation case is checked?
```
It doesn't seem like it? I just used Hyper Beam a bunch (gave the other mon Attract) and it never triggered infatuation on the recharge turn. I guess it's possible I failed ~10 coinflips though.

flx-sta commented on src/field/arena.ts:
```diff
@@ -391,16 +391,16 @@ export class Arena {
-  isMoveTerrainCancelled(user: Pokemon, targets: BattlerIndex[], move: Move) {
-    return this.terrain && this.terrain.isMoveTerrainCancelled(user, targets, move);
+  public isMoveTerrainCancelled(user: Pokemon, targets: BattlerIndex[], move: Move): boolean {
+    return (this.terrain && this.terrain.isMoveTerrainCancelled(user, targets, move)) as boolean;
```
You should be able to avoid `as boolean` by using `!!this.terrain`.
<sub>This is caused by the fact that you are checking this.terrain in a return and not in a if-clause. The compiler doesn't automatically know that you are simply checking if it's defined</sub>

```suggestion
    return !!this.terrain && this.terrain.isMoveTerrainCancelled(user, targets, move);
```

---

Alternatively you could also simply use `?.`

```suggestion
    return this.terrain?.isMoveTerrainCancelled(user, targets, move);
```

flx-sta commented on src/field/arena.ts:
```diff
@@ -391,16 +391,16 @@ export class Arena {
-  isMoveWeatherCancelled(user: Pokemon, move: Move) {
-    return this.weather && !this.weather.isEffectSuppressed(this.scene) && this.weather.isMoveWeatherCancelled(user, move);
+  public isMoveWeatherCancelled(user: Pokemon, move: Move): boolean {
+    return (this.weather && !this.weather.isEffectSuppressed(this.scene) && this.weather.isMoveWeatherCancelled(user, move)) as boolean;
```
You should be able to avoid `as boolean` by using `!!this.weather`.
<sub>This is caused by the fact that you are checking this.terrain in a return and not in a if-clause. The compiler doesn't automatically know that you are simply checking if it's defined</sub>

```suggestion
    return !!this.weather && !this.weather.isEffectSuppressed(this.scene) && this.weather.isMoveWeatherCancelled(user, move)) as boolean;
```

---

Alternatively you could also simplify this by using `.?`

```suggestion
    return !this.weather?.isEffectSuppressed(this.scene) && this.weather.isMoveWeatherCancelled(user, move);
```
<sup> Only the first one needs it as the it's an AND connection and the 2nd one will be guaranteed for it to be set</sup>

DayKev commented on src/phases/move-phase.ts:
```diff
@@ -273,25 +186,257 @@ export class MovePhase extends BattlePhase {
+
     } else {
-      doMove();
+      if ([Moves.ROAR, Moves.WHIRLWIND, Moves.TRICK_OR_TREAT, Moves.FORESTS_CURSE].includes(this.move.moveId)) {
```
```suggestion
      if ([ Moves.ROAR, Moves.WHIRLWIND, Moves.TRICK_OR_TREAT, Moves.FORESTS_CURSE ].includes(this.move.moveId)) {
```
